chore(kuma-cp): log tenantID + tracing's traceID and spanID if present

bartsmykla opened the pull request:
Log tenantID + tracing's traceID and spanID if present

### Checklist prior to review

<!--
Each of these sections need to be filled by the author when opening the PR.

If something doesn't apply please check the box and add a justification after the `--`
-->

- [x] [Link to relevant issue][1] as well as docs and UI issues
  - No relevant issue
- [x] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS
  - It won't
- [x] Tests (Unit test, E2E tests, manual test on universal and k8s)
  - Tests were updated
- [x] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)?
  - There is no need
- [x] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? 
  - There is no need
- [x] Do you need to explicitly set a [`> Changelog:` entry here](../blob/master/CONTRIBUTING.md#submitting-a-patch) or add a `ci/` label to run fewer/more tests?

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

slonka reviewed (commented):
I think it's good but I'm not sure about 2 things:
1. When can a logger be missing from the context? It's not a huge deal but having to do error checking every time you need a logger seems like a lot of boilerplate. Maybe we can remove some default global one in that case?
2. Do we need all the variations of `FromContextWithXAndY`?

michaelbeaumont commented:
>  When can a logger be missing from the context? It's not a huge deal but having to do error checking every time you need a logger seems like a lot of boilerplate. Maybe we can remove some default global one in that case?

:+1: on this

> Do we need all the variations of FromContextWithXAndY?

Agree, why not use `logger.Values` etc instead of `WithNameAndValues`?

lahabana commented:
+1 to the fallback. For .e.g you always pass as a default: core.Log. You could always add a debug log to mention there's no logger in the context if we have fun chasing missing logger in context.

The other thing I'm slightly confused about is why we need the logger in the context. In other words: is there a reason for the logger to be different on a request basis? I believe the answer is no and only metadata is different. So I feel what we should have is a `logger.WithCtx(ctx) Logger` or `Logger.InfoCtx(ctx, msg, kv...)` rather than adding a logger to the context when it's always the same. WDYT?

bartsmykla commented:
> The other thing I'm slightly confused about is why we need the logger in the context. In other words: is there a reason for the logger to be different on a request basis? I believe the answer is no and only metadata is different. So I feel what we should have is a `logger.WithCtx(ctx) Logger` or `Logger.InfoCtx(ctx, msg, kv...)` rather than adding a logger to the context when it's always the same. WDYT?

It makes sense, the problem is we are using `logr.Logger` everywhere, and that would involve us introducing some new abstraction (we could even use slog then) which I would say should be done separately.

bartsmykla commented:
> When can a logger be missing from the context? It's not a huge deal but having to do error checking every time you need a logger seems like a lot of boilerplate. Maybe we can remove some default global one in that case?

+1 to the fallback. For .e.g you always pass as a default: core.Log. You could always add a debug log to mention there's no logger in the context if we have fun chasing missing logger in context.

addressed, but had to extract these functions to separate directory/package to not introduce cyclic dependency

> > Do we need all the variations of FromContextWithXAndY?
> 
> Agree, why not use `logger.Values` etc instead of `WithNameAndValues`?

addressed

lahabana commented on pkg/log/context/context.go:
```diff
@@ -0,0 +1,23 @@
+package context
```
Instead of this move the global `Log` inside the logger package and reference `Log = logger.DefaultLog`. This should avoid the cycling dep and keep packages simple no?

bartsmykla replied:
addressed

lahabana reviewed (commented):
Ok I wrote: `logger.WithCtx(ctx) Logger` but I actually meant it as a utility method: `logger.DecorateWithCtx(logger, ctx) Logger`

bartsmykla commented:
> Ok I wrote: `logger.WithCtx(ctx) Logger` but I actually meant it as a utility method: `logger.DecorateWithCtx(logger, ctx) Logger`

I updated the PR to follow your suggestion

lahabana commented on pkg/log/logger.go:
```diff
@@ -86,3 +88,38 @@ func newZapLoggerTo(destWriter io.Writer, level LogLevel, opts ...zap.Option) *z
+	fields []interface{}
+}
+
+func WithFields(ctx context.Context, keysAndValues ...interface{}) context.Context {
```
Is the idea to call this in `multitenant.WithTenant` ? If so maybe you should do it now to make this PR more understandable. In any case it's worth adding comments to both these functions

bartsmykla replied:
addressed

jakubdyszkiewicz commented on pkg/multitenant/multitenant.go:
```diff
@@ -31,7 +38,14 @@ func (s singleTenant) GetIDs(context.Context) ([]string, error) {
+		span.SetAttributes(attribute.String(tracingAttributeName, tenantId))
+	}
+
+	return kuma_log.WithFields(
```
Can we name this function `CtxWithFields` or `CtxWithLogFields`? When I saw `kuma_log.WithFields` my guess was that we return a logger with fields (since the package is about logging), but that's not true.

jakubdyszkiewicz commented on pkg/log/logger.go:
```diff
@@ -86,3 +88,39 @@ func newZapLoggerTo(destWriter io.Writer, level LogLevel, opts ...zap.Option) *z
+	if span := trace.SpanFromContext(ctx); span.IsRecording() {
+		logger = logger.WithValues(
+			"trace_id", span.SpanContext().TraceID(),
+			"span_id", span.SpanContext().SpanID(),
```
nit but I'll ask anyways. Our convention for keys in log is camelCase. Is there a reason you use snake_case here?

bartsmykla replied:
Yes, that's the property name used to extract trace ids by datadog, dynatrace etc.

lahabana commented on pkg/multitenant/multitenant.go:
```diff
@@ -31,7 +38,14 @@ func (s singleTenant) GetIDs(context.Context) ([]string, error) {
-	return context.WithValue(ctx, tenantCtx{}, tenantId)
+	if span := trace.SpanFromContext(ctx); span.IsRecording() {
+		span.SetAttributes(attribute.String(tracingAttributeName, tenantId))
+	}
```
Ho sweet tenant in traces!

bartsmykla replied:
😎

lahabana commented on pkg/multitenant/multitenant.go:
```diff
@@ -3,8 +3,15 @@ package multitenant
+	kuma_log "github.com/kumahq/kuma/pkg/log"
 )
 
+const tracingAttributeName = "tenantID"
```
```suggestion
const observabilityAttributeName = "tenantID"
```

bartsmykla replied:
addressed

michaelbeaumont commented on pkg/log/logger.go:
```diff
@@ -86,3 +88,39 @@ func newZapLoggerTo(destWriter io.Writer, level LogLevel, opts ...zap.Option) *z
+// CtxWithFields will add to provided context fields (key/value pairs)
+// which then will be logged by logger decorated with DecorateWithCtx function.
+// After adding logging fields it returns back enriched context.
+func CtxWithFields(ctx context.Context, keysAndValues ...interface{}) context.Context {
```
why not just use `.Values`? Why include values in this function?

bartsmykla replied:
We've followed up [here](https://github.com/kumahq/kuma/pull/7285)
